Ble remove user facing abstraction

pan- opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

Completion of BLE streamlining: 
- Makes public API relevant and not just fake for doxygen
- Reorganise public headers:
  - `ble/common`: For code common in gatt, gap and drivers
  - `ble/driver`: HCI driver interface declaration
  - `ble/gap`: declarations related to gap
  - `ble/gatt`: declarations related to gatt
  - `ble/service`: ble services
  - `ble/compatibility`: Legacy path for backward compatibility 
- reorganise sources and move private headers with them: 
  - `source`: Entry point to private implementation of Gap, GattClient, GattServer and SecurityManager
  - `source/generic`: Generic implementation for Gap, GattClient, GattServer and SecurityManager
  - `source/pal`: PAL declaration used by the generic implementation
  - `source/gatt`: Definition of public gatt related classes
  - `source/gap`: Definition of public gap related classes
  - `source/cordio`: Implementation of the pal and assembly of the various elements in BleInstanceBaseImpl. 
- move cordio stack out of `TARGET_CORDIO`. It is located in `ble/libraries`
- Expose HCI driver at the root of BLE API. 


#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

@paul-szczepanek-arm @LDong-Arm 

----------------------------------------------------------------------------------------------------------------

pan- committed:
PIMPL for ble::Gap

pan- committed:
ble: GattClient PIMPL

pan- committed:
BLE: SecurityManager PIMPL

pan- committed:
BLE: GattServer PIMPL

pan- committed:
BLE: Move constructor from transport to private section

pan- committed:
BLE: Bind interface to private implementation.

pan- committed:
BLE: Warning cleanup

- Missing default assignment operator
- Field order in construction list
- Typo
- Unused header

pan- committed:
BLE: Move generic implementation and headers into source/generic

pan- committed:
BLE: Move SecurityDB related files into generic sources

pan- committed:
BLE: Move BleInstanceBase.h into source.

pan- committed:
BLE: move PAL headers into source/pal

pan- committed:
BLE: Move gap public header at the root.

pan- committed:
BLE: Move common publi headers into ble/common

pan- committed:
BLE: Move GATT public header into ble/gatt

Move private headers GattServerEvents.h, GapEvents,h and GapType.h in relevant part of the source tree.

pan- committed:
BLE: Add compatibility path for headers.

pan- committed:
BLE: Fix address copy in Nordic HCI driver.

pan- committed:
BLE: Move cordio_stack out of TARGET_CORDIO

pan- committed:
BLE: Move generic lib file into source/generic .

pan- committed:
BLE: Move Cordio implementation in source/impl

pan- committed:
BLE: Expose HCI driver in public API.

pan- committed:
BLE: Clang tidy on public headers.

pan- committed:
BLE: Cleanup public headers

pan- committed:
BLE: move source/impl into source/cordio

pan- committed:
BLE: Move source/DiscoveredCharacteristic into source/gatt/DiscoveredCharacteristic

pan- committed:
BLE: cleanup generic sources

pan- committed:
BLE: Cleanup of pal AttClient and GattClient

Fix interface namespace being not used.
Move generic AttClient to GattClient adapter back into the  pal

pan- committed:
BLE: Move GattServer implementation from generic to cordio

pan- committed:
BLE: Cleanup pal Gap

pan- committed:
BLE: Cleanup pal security manager

pan- committed:
BLE: cleanup PalEvent queue and implementation

pan- committed:
BLE: Cleanup pal generic access service

pan- committed:
BLE: Cleanup PalSigningEventMonitor.

pan- committed:
BLE: Remove cordio namespace

pan- committed:
BLE: Cleanup AttServerMessage implementation

pan- committed:
BLE: Cleanup BLEInstanceBas implementation

pan- committed:
BLE: Move cordio implementation include into source folder

pan- committed:
BLE: update doxygen exclusion pattern

pan- committed:
BLE: Fix uses of mbed.h

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/include/ble/common/Bounded.h:
```diff
@@ -47,9 +47,9 @@ struct Bounded {
     {
         if (v < Min) {
-            _value = v;
+            _value = Min;
```
this is unexpected in a "refactor"

pan- replied:
Ahah, little bug I caught, it should have had its own commit.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/cordio/source/BLEInstanceBaseImpl.cpp:
```diff
@@ -17,9 +17,26 @@
-#include "CriticalSectionLock.h"
+#include "ble/driver/CordioHCIDriver.h"
+
+#include "source/pal/PalAttClient.h"
```
why are headers in the source folder?

pan- replied:
Because there is private headers and public headers. public headers goes into `include/<module name>` while private headers goes into `source`

paul-szczepanek-arm replied:
that's not how we agreed to do it, private headers go into internal, did you ask core about this?

pan- replied:
Yes I did ask Core about this. Obviously one of the precondition for this is that private headers are **not** included by public ones which wasn't possible with the previous separation. 
The internal _trick_ is used when there is private declarations used by public headers we put them into `include/internal` so users know they should not use it.
There's no such header, anymore and with the Make-ification these private headers won't be visible to consumers of BLE.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/cordio/source/BLEInstanceBaseImpl.cpp:
```diff
@@ -369,21 +373,21 @@ void BLEInstanceBase::stack_handler(wsfEventMask_t event, wsfMsgHdr_t *msg)
-
-                _hci_driver->handle_test_end(status == 0, packet_number);
-                return;
+            case DM_UNHANDLED_CMD_CMPL_EVT_IND: {
```
wrong indent

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/include/ble/Gap.h:
```diff
@@ -1435,29 +1431,36 @@ class Gap {
+    /*
+     * Constructor from the private implementation.
+     */
+    Gap(impl::Gap* impl) : impl(impl) {}
```
I'm fine with this but I want to be on record being against an extra step of indirection, I don't believe making it look neater is worth the cost

pan- replied:
There are other options such as implementing things directly and using the private pointer for holding data but that was too much changes in one go. Private implementation makes us more robust regards to ABI changes and cleanly isolate from private headers which won't be accessible with the CMakeification.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/include/ble/GattClient.h:
```diff
@@ -625,7 +625,10 @@ class GattClient {
-    void onShutdown(T *objPtr, void (T::*memberPtr)(const GattClient *));
+    void onShutdown(T *objPtr, void (T::*memberPtr)(const GattClient *))
+    {
+        onShutdown({objPtr, memberPtr});
```
why is this not in implementation?

pan- replied:
Because it is a template.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/include/ble/gap/AdvertisingDataSimpleBuilder.h:
```diff
@@ -68,7 +68,7 @@ class AdvertisingDataSimpleBuilder {
      * Construct a AdvertisingDataSimpleBuilder
      */
-    AdvertisingDataSimpleBuilder() : _builder(_buffer)
+    AdvertisingDataSimpleBuilder() : _builder(_buffer), _buffer()
```
wrong order of init

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/include/ble/services/BatteryService.h:
```diff
@@ -86,7 +86,7 @@ class BatteryService {
             GattService::UUID_BATTERY_SERVICE,
             charTable,
-            sizeof(charTable) / sizeof(GattCharacteristic *)
+            sizeof(charTable) / sizeof(charTable[0])
```
~why? that is less readable than just size of pointer which is what it contains~
actually, I think it's fine

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/pal/PalAttClientToGattClient.cpp:
```diff
@@ -16,31 +16,32 @@
 namespace ble {
 
-PalGattClient::PalGattClient(PalAttClient& client) : _event_handler(nullptr), _client(client)
+PalAttClientToGattClient::PalAttClientToGattClient(PalAttClient& client) : _event_handler(nullptr), _client(client)
```
why the name change? is this not the PalGattClient? Is it relevant that it uses AttClient internally?

pan- replied:
No, it is not the `PalGattClient` interface, this is a generic **implementation** of the `PalGattClient` interface using a `PalAttClient`. I moved it outside of Cordio as it is independent from it.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/SecurityManagerImpl.h:
```diff
@@ -19,62 +19,74 @@
+    ble_error_t setDatabaseFilepath(const char *dbFilepath = nullptr);
 
-    ble_error_t reset(void);
+    ble_error_t reset(ble::SecurityManager* sm);
```
why is secruity manager passed in to a non static function of security manger?

pan- replied:
The body invoke a callbacks that needs a pointer to the `SecurityManager` seen by the user.

paul-szczepanek-arm commented on connectivity/FEATURE_BLE/source/generic/GattClientImpl.h:
```diff
@@ -222,21 +243,29 @@ class GattClient :
+    mutable ProcedureControlBlock *control_blocks;
     bool _is_reseting;
 
+    ::ble::GattClient *client = nullptr;
```
why would the implementation need a pointer to the interface it's implementing?

pan- replied:
The class outputs object that reference the `GattClient` interface. 
I will fix both the GattClient and SecurityManager so it uses the BLE singleton instead.

paul-szczepanek-arm replied:
Sounds good, maybe fix the resets as well

paul-szczepanek-arm reviewed (changes requested):
Great work on fixing all the astyle and adding autio, override, final etc. and sorting out headers. I think the extra layer of indirection is not worth it and complicates the implementation but that's just my opinion.

The only change needed is the private headers should be in internal according to this plan: https://confluence.arm.com/display/IoTBU/IOTCLOUDPR-3432+Clean-up+Mbed+OS+directory+structure

paul-szczepanek-arm reviewed (approved):
OK, if core are fine with the the new header structure I have nothing against it.

pan- committed:
BLE: Fixed used of parent pointer in implementation classes.

pan- committed:
BLE: correctness improvments

Remove unused variable, improve header inclusion, fix initialization list order,

pan- committed:
BLE: fix code indentation

pan- committed:
BLE: reintroduce default implementation GattClient::discoverServices.

pan- committed:
Cypress BLE driver: Fix removal of mbed.h

0xc0170 reviewed (approved):
We might need your help to get this in also to the feature-cmake branch.
